src/charm.py

```diff
@@ -303,20 +328,180 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
+        except ExecError as err:
+            logger.error(f"Exited with code: {err.exit_code}. Stderr: {err.stderr}")
+            return
+        except Error as e:
+            logger.error(f"Something went wrong when trying to run the command: {e}")
+            logger.info("Deferring the event")
+            event.defer()
+            return
+
+    def _update_endpoint_info(self) -> None:
+        if not self.admin_ingress.url or not self.public_ingress.url:
+            return
+
+        self.oauth.set_provider_info_in_relation_data(
+            issuer_url=self.public_ingress.url,
+            authorization_endpoint=join(self.public_ingress.url, "oauth2/auth"),
+            token_endpoint=join(self.public_ingress.url, "oauth2/token"),
+            introspection_endpoint=join(self.admin_ingress.url, "admin/oauth2/introspect"),
+            userinfo_endpoint=join(self.public_ingress.url, "userinfo"),
+            jwks_endpoint=join(self.public_ingress.url, ".well-known/jwks.json"),
+            scope=SUPPORTED_SCOPES,
+        )
+
+
+class HydraCLI:
+    """Helper object for running hydra CLI commands."""
+
+    def __init__(self, hydra_admin_url: str, container: Container):
+        self.hydra_admin_url = hydra_admin_url
+        self.container = container
+
+    def _client_config_to_cmd(
+        self, client_config: ClientConfig, metadata: Optional[Union[Dict, str]] = None
+    ) -> List[str]:
+        """Convert a ClientConfig object to a list of parameters."""
+        flags = [
+            "--grant-type",
+            ",".join(client_config.grant_types or ["authorization_code", "refresh_token"]),
+            "--response-type",
+            "code",
+        ]
+
+        if client_config.scope:
+            for s in client_config.scope.split(" "):
+                flags.append("--scope")
+                flags.append(s)
+        if client_config.redirect_uri:
+            flags.append("--redirect-uri")
+            flags.append(client_config.redirect_uri)
+        if metadata:
+            flags.append("--metadata")
+            if isinstance(metadata, dict):
+                flags.append(json.dumps(metadata))
+            elif isinstance(metadata, str):
+                flags.append(metadata)
+            else:
+                raise ValueError("Invalid metadata")
+        return flags
+
+    def _client_cmd_prefix(self, action: str) -> List[str]:
+        return [
+            "hydra",
+            action,
+            "client",
+            "--endpoint",
+            self.hydra_admin_url,
+            "--format",
+            "json",
+        ]
+
+    def create_client(
+        self, client_config: ClientConfig, metadata: Optional[Union[Dict, str]] = None
+    ) -> Optional[Dict]:
+        """Create an oauth2 client."""
+        cmd = self._client_cmd_prefix("create") + self._client_config_to_cmd(
+            client_config, metadata
+        )
+
+        stdout, _ = self._run_cmd(cmd)
+        stdout = json.loads(stdout)
```

Comment:
I'd use a different variable, the type/content is different from the previous `stdout`